### PR TITLE
Stats

### PR DESCRIPTION
**Pull request by pernas**

### CODE REVIEW

**Comment by Sjors** [comment]
Looks like you made linter unhappy:
```
  861:4  error  Missing semicolon  semi
  868:2  error  Missing semicolon  semi
```

File: src/blockchain-wallet.js [hunk 1]
```diff
@@ -854,3 +854,15 @@ Wallet.prototype.loadExternal = function () {
...
+
+Wallet.prototype.incStats = function () {
+  var balanceNoWO = function (a) {
+    return a.isWatchOnly ? 0 : a.balance;
+  }
+  var legacyBalance = this.keys.map(balanceNoWO).reduce(Helpers.add, 0);
+  var mvBool = this.hdwallet ? this.hdwallet.isMnemonicVerified : false;
+  API.incrementSecPassStats(this.isDoubleEncrypted);
+  API.incrementRecoveryStats(mvBool);
+  API.incrementLegacyUseStats(legacyBalance > 0);
```
Comment (Sjors) [inline src/blockchain-wallet.js]: Maybe use a slightly higher threshold, or filter for dust in each address?

Comment (pernas) [inline src/blockchain-wallet.js]: I don't think the dust is important. Is for stats collection. It is ok if we just distinct never used addresses a grosso modo

File: src/wallet.js [hunk 2]
```diff
@@ -353,8 +353,14 @@ MyWallet.initializeWallet = function (pw, decryptSuccess, buildHdSuccess) {
     };
     return MyWallet.wallet.loadExternal.bind(MyWallet.wallet)().catch(loadExternalFailed);
   };
-
-  return Promise.resolve().then(doInitialize).then(tryLoadExternal);
+  var incStats = function () {
+    var gh = MyWallet.wallet.getHistory.bind(MyWallet.wallet);
```
Comment (pernas) [inline src/wallet.js]: Yeah I though about this... The problem is the getHistory calls on mywallet v3 are used ignoring the returned promise. So, I cannot use them without some big refactoring in the login process and I don't want to split this collection data features into my-wallet-v3 and frontend...

Comment (Sjors) [inline src/wallet.js]: We can at least make this `getHistory` call more efficient:
1 - only ask the balance, not the full transaction history
2 - only check legacy address
3 - don't check watch-only addresses

Comment (Sjors) [inline src/wallet.js]: I agree we should not involve the frontend.

File: src/blockchain-wallet.js (hunk 1 above)
Comment (Sjors) [inline src/blockchain-wallet.js]: `incrementLegacyUseStats` only sends a boolean, not the balance. So there would be a lot of false positives for how many people use legacy.

Comment (Sjors) [inline src/blockchain-wallet.js]: But we can change the threshold in a later version of course.

**Comment by Sjors** [comment]
One problem with always using POST to submit the guid is that every login the console will show this:
schermafbeelding 2016-10-12 om 09 10 18

I also suspect a GET request can be more efficiently cached.

File: src/wallet.js [hunk 3]
```diff
@@ -353,8 +353,17 @@ MyWallet.initializeWallet = function (pw, decryptSuccess, buildHdSuccess) {
...
+    var is = MyWallet.wallet.incStats.bind(MyWallet.wallet);
+    return gh().then(is);
+  };
+  var saveGUID = function () {
+    return MyWallet.wallet.saveGUIDtoMetadata();
+  };
+  var p = Promise.resolve().then(doInitialize);
+  p.then(incStats).then(saveGUID);
```
Comment (pernas) [inline src/wallet.js]: 1) yes
2) yes

**Comment by pernas** [comment]
yes, I could catch it to hide the error.. but we are not going to do anything with the catch

**Comment by Sjors** [comment]
Catching a 404 error won't make it go away in the console. Why not just do a fetch and it that fails, do a create?

**Comment by pernas** [comment]
yes we could, but what is the problem of having that error? It is something expected. I want to save something if it was never saved before. And this is what post is doing

**Comment by Sjors** [comment]
Too many meaningless error messages in the console make it more difficult to spot the important ones. 

All other metadata features first make a GET request to see if a record exists and then make a POST request to create a new one, or a PUT request to update the existing one. What is the benefit of doing it different here, except for saving 1 line of code?

**Comment by Sjors** [comment]
I added some tests, and I made it do a `fetch()` first. 

I'll look into making the wallet balance fetch call a bit more efficient...

**Comment by Sjors** [comment]
I replaced `getHistory` with `getBalances` and only put legacy, not watch-only, and  non archived addresses in the request. Also added more tests.

**Comment by pernas** [comment]
closed in favor of https://github.com/blockchain/My-Wallet-V3/pull/311

**Pull request closed without merging**